### PR TITLE
Upgrade to node 10 and increase memory usage

### PR DESCRIPTION
**Pull request by npeternel**
Moving from Node4 to Node10 to match the Node version in ecs-watchbot@4.20.0

This PR also uses a soft limit on memory and increases the initial reservation to account for larger binary file sizes from the version upgrade. This prevents exit code 137 errors we saw while testing out the new version.

### CODE REVIEW

**Comment by rclark** [comment]
**tl;dr**

- I can't find evidence that memory consumption caused any watchbot containers to be killed.
- Exit code 137 should be expected from watchbot containers, because they don't listen to SIGTERM.
- That's not the same thing as an exit code 137 printed in watchbot application logs. These logs indicate that the worker process (ie what to do with a message, defined by ecs-telephone) was killed.
- Container insights, Athena, and CloudWatch logs are a powerful set of tools.

---

> Moving from Node4 to Node10 to match the Node version in ecs-watchbot@4.20.0

Technically, there is no requirement that these versions align. Watchbot is shipped as a binary that includes its own node.js version (10). When the container starts, that binary runs. It collects messages from SQS and when it receives one, it launches the "worker" as a child process. In the case of ecs-telephone, that worker process was a node.js script that was supposedly running on node.js 4. But it could be anything you want it to be. Python, Go, Node.js 12, w/e.

> This PR also uses a soft limit on memory and increases the initial reservation to account for larger binary file sizes from the version upgrade. This prevents exit code 137 errors we saw while testing out the new version.

Has this been observed to solve the problem reported in https://github.com/mapbox/ecs-watchbot/issues/327? 

I looked at memory utilized by task over the last week via container insights, looking for any containers with high memory utilization.

```
fields ServiceName, MemoryUtilized, @message
| filter Type = 'Task'
| filter @message like /ecs-telephone/
| sort by MemoryUtilized desc
```

These data are collected once per minute per task. In almost every case, MemoryUtilized was `64`, but there was one datapoint at `111` for task `cfba54028d3d4871a75ee33fba73ca89`.

I turned to Athena to observe the lifecycle of that Task. The `production_loading_dock.ecs_state_production_tasks` table records every lifecycle event for every ECS task.

```
select time,
       json_extract_scalar(detail, '$.lastStatus') as status,
       json_extract_scalar(detail, '$.containers[0].exitCode') as code,
       json_extract_scalar(detail, '$.stoppedReason') as reason

from production_loading_dock.ecs_state_production_tasks

where dt = '2019-12-24'
      and json_extract_scalar(detail, '$.taskArn') like '%cfba54028d3d4871a75ee33fba73ca89'

order by time asc
```

The result indicates that this task was stopped due to ECS service scaling behavior, not killed by memory consumption.

image

I believe that this exit code can indicate that the container received a signal from ECS to stop (SIGTERM), but ignored that signal. Subsequently, ECS would send a SIGKILL signal.

> This results in a SIGTERM value and a default 30-second timeout, after which the SIGKILL value is sent and the containers are forcibly stopped. If the container handles the SIGTERM value gracefully and exits within 30 seconds from receiving it, no SIGKILL value is sent.

Confused, I took a closer look at some application logs from watchbot that include the `Killed` message reported in https://github.com/mapbox/ecs-watchbot/issues/327. In such a task (`cfdbf35d84e34dd3a55be0c8a98bfaf6`) the logs loop like this:

```
[Fri, 20 Dec 2019 01:24:18 GMT] [watcher] [2f8cf7c8-82d9-4f6f-b154-b5ecab39f0fc] 
{
    "message": "\"hello hello helllo\"",
    "receives": "1"
}
[Fri, 20 Dec 2019 01:24:19 GMT] [worker] [2f8cf7c8-82d9-4f6f-b154-b5ecab39f0fc] Killed
[Fri, 20 Dec 2019 01:24:19 GMT] [watcher] [2f8cf7c8-82d9-4f6f-b154-b5ecab39f0fc] [failure] 
{
    "code": 137,
    "signal": null,
    "duration": 891,
    "response_duration": 902
}
[Fri, 20 Dec 2019 01:24:21 GMT] [watcher] [2f8cf7c8-82d9-4f6f-b154-b5ecab39f0fc] 
{
    "message": "\"hello hello helllo\"",
    "receives": "2"
}
```

It receives a message, launches the child process. Then `[worker] ... Killed` indicates an action taken by the host OS on the child process. The job that watchbot launched was killed, but the watchbot process itself was not. The watchbot process returns the message to SQS and then picks it up again.

I took this new task's ID back to container insights to look at the memory utilization over time, and it appears that it never reported usage higher than 64 MiB:

```
fields ServiceName, MemoryUtilized, @message
| filter Type = 'Task'
| filter @message like /ecs-telephone/
| filter TaskId = 'cfdbf35d84e34dd3a55be0c8a98bfaf6'
| stats max(MemoryUtilized) by bin(10m)
```

image

The task ran from 12/20 - 12/24 on this ecs-telephone commit https://github.com/mapbox/ecs-telephone/commit/b54ef3a1ad81ee7ff8f1ceb89992ca06bb2530f5 (the commit sha shows up on the cloudwatch log stream's prefix).

Back to Athena to see what this tasks lifecycle looked like:

```
select time,
       json_extract_scalar(detail, '$.lastStatus') as status,
       json_extract_scalar(detail, '$.containers[0].exitCode') as code,
       json_extract_scalar(detail, '$.stoppedReason') as reason

from production_loading_dock.ecs_state_production_tasks

where dt > '2019-12-19'
      and dt < '2019-12-25'
      and json_extract_scalar(detail, '$.taskArn') like '%cfdbf35d84e34dd3a55be0c8a98bfaf6'

order by time asc
```

image

This task was killed when the EC2 it was running on was terminated for one reason or another. Again, not due to any kind of problematic memory usage.
